### PR TITLE
Cutadapt process

### PR DESCRIPTION
**Pull request by jvrakor**
@jkokosar @JureZmrzlikar @mstajdohar @dblenkus

### CODE REVIEW

File: resolwe_bio/processes/reads_processing/cutadatp.yml [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+#
+# Processors for NGS reads filtering/trimming
+---
+
+- slug: cutadapt-single
+  name: Cutadapt (single end)
+  requirements:
+    expression-engine: jinja
+  data_name: "Cutadapt ({{ reads|sample_name|default('?') }})"
+  version: 0.0.0
```
Comment (jkokosar) [inline resolwe_bio/processes/reads_processing/cutadatp.yml]: version 0.0.1

File: resolwe_bio/processes/reads_processing/cutadatp.yml [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    - name: down_primers_seq
+      label: 3 prime adapter sequence
+      type: basic:string
+      required: false  
+    - name: polya_tail
+      label: Poly-A tail
+      type: basic:integer
+      required: false
+      description: >
+        Lenght of poly-A tail, example - AAA -> 3, AAAAA -> 5
```
Comment (jkokosar) [inline resolwe_bio/processes/reads_processing/cutadatp.yml]: Length

File: resolwe_bio/processes/reads_processing/cutadatp.yml [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+      type: list:basic:file:html
+    - name: fastqc_archive
+      label: Download FastQC archive
+      type: list:basic:file
+  run:
+    runtime: polyglot
+    language: bash
+    program: |
+
+      {% if not up_primers_file and not down_primers_file and not up_primers_seq and not down_primers_seq
```
Comment (jkokosar) [inline resolwe_bio/processes/reads_processing/cutadatp.yml]: Split this into multiple lines using `\`

File: resolwe_bio/processes/reads_processing/cutadatp.yml [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      required: false
+      description: >
+        Drop the read if it is below a specified length   
+  output:
+    - name: fastq
+      label: Reads file
+      type: list:basic:file
+    - name: report
+      label: Cutadapt report
+      type: list:basic:file
```
Comment (jkokosar) [inline resolwe_bio/processes/reads_processing/cutadatp.yml]: is this really a list?

File: resolwe_bio/processes/reads_processing/cutadatp.yml [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+      {% if up_primers_file %} cutadapt -O {{min_overlap_up}} -g file:5_primers.fasta - 2>> report.txt | {% endif %}
+      {% if minlen %} cutadapt -m {{minlen}} - 2>> report.txt | {% endif %}
+      cat - > ${NAME}_trimmed.fastq
+      re-checkrc "Failed while running cutadapt"
+      re-progress 0.7
+
+      gzip ${NAME}_trimmed.fastq
+
+      re-save-file-list fastq ${NAME}_trimmed.fastq.gz
+      re-save-file-list report report.txt
```
Comment (jkokosar) [inline resolwe_bio/processes/reads_processing/cutadatp.yml]: see above

File: resolwe_bio/processes/reads_processing/cutadatp.yml [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+      {% endif %}
+
+      {% if up_primers_file and up_primers_seq %}
+        re-error "5 prime adapters should be either a sequence or in a .fasta file, not both"
+      {% endif %}
+
+      NAME=`basename {{reads.fastq.0.file}} .fastq.gz`
+
+      READS={{ reads.fastq|join(' ', attribute='file') }}
+      cat "${READS}" > in_reads.fastq.gz
```
Comment (jkokosar) [inline resolwe_bio/processes/reads_processing/cutadatp.yml]: "${READS[@]}" to concat a list

File: resolwe_bio/tests/processes/test_reads_filtering.py [hunk 7]
```diff
@@ -157,6 +157,43 @@ def test_hsqutils_trimm(self):
         self.assertFiles(hsqutils_trimm, 'fastq2', ['hsqutils_reads_trimmed_mate2.fastq.gz'],
                          compression='gzip')
 
+    def test_cutadapt(self):
+        self.keep_data()
```
Comment (jkokosar) [inline resolwe_bio/tests/processes/test_reads_filtering.py]: remove this

File: resolwe_bio/tests/processes/test_reads_filtering.py [hunk 8]
```diff
@@ -157,6 +157,43 @@ def test_hsqutils_trimm(self):
...
+        inputs = { 'src': ['cutadapt_single.fastq.gz']}
```
Comment (jkokosar) [inline resolwe_bio/tests/processes/test_reads_filtering.py]: in test, duplicate the file here to test "list" functionality.

File: resolwe_bio/processes/reads_processing/cutadatp.yml (hunk 5 above)
Comment (jvrakor) [inline resolwe_bio/processes/reads_processing/cutadatp.yml]: To what does this comment apply?

File: resolwe_bio/processes/reads_processing/cutadatp.yml [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+  requirements:
+    expression-engine: jinja
+  data_name: "Cutadapt ({{ reads|sample_name|default('?') }})"
+  version: 0.0.1
+  type: data:reads:fastq:single:cutadapt
+  category: analyses
+  flow_collection: sample
+  persistence: CACHED
+  description: >
+    "Trimm 5' and 3' primers using Cutadapt"
```
Comment (mzganec) [inline resolwe_bio/processes/reads_processing/cutadatp.yml]: Trimm -> Trim

File: resolwe_bio/tests/processes/test_reads_filtering.py (hunk 7 above)
Comment (jvrakor) [inline resolwe_bio/tests/processes/test_reads_filtering.py]: Will remove when everything else is ok

File: resolwe_bio/tests/processes/test_reads_filtering.py [hunk 10]
```diff
@@ -157,6 +157,91 @@ def test_hsqutils_trimm(self):
         self.assertFiles(hsqutils_trimm, 'fastq2', ['hsqutils_reads_trimmed_mate2.fastq.gz'],
                          compression='gzip')
 
+    def test_cutadapt_single(self):
+        inputs = {'src': ['cutadapt_single.fastq.gz', 'cutadapt_single1.fastq.gz']}
+        reads = self.run_process('upload-fastq-single', inputs)
```
Comment (mzganec) [inline resolwe_bio/tests/processes/test_reads_filtering.py]: Use `prepare_reads(['cutadapt_single.fastq.gz', 'cutadapt_single1.fastq.gz'])`, similarly in other places.

File: resolwe_bio/tests/processes/test_reads_filtering.py [hunk 11]
```diff
@@ -157,6 +157,91 @@ def test_hsqutils_trimm(self):
...
+            'minlen': 10
+        }
+
+        cutadapt_single = self.run_process('cutadapt-single', inputs)
+
+        self.assertFiles(cutadapt_single, 'fastq', ['cutadapt_single_trimmed.fastq.gz'],
+                         compression='gzip')
+
+        inputs = {'src': '5_prime_adapter.fasta.gz'}
+        primers_up = self.run_processor('upload-fasta-nucl', inputs)
```
Comment (mzganec) [inline resolwe_bio/tests/processes/test_reads_filtering.py]: Use `prepare_adapters('5_prime_adapter.fasta.gz')`, similarly in other places.

**Pull request merged**